### PR TITLE
Add algorithm CMA-ME

### PR DESCRIPTION
**Pull request by felixchalumeau**
Related to issue #80 

Done:
- implement the improvment emitter
- implement the random emitter
- implement the opt emitter
- implement the pool of emitters
- example notebook
- check perfs of the CMA ES optimizer
- fix bug in CMA-MEGA (issue in improvment computation due to bad repertoire use)
- getting similar perfs for the improv emitter on ragistrin and sphere
- even outperforming with other params
- update documentation
- add tests

Future improvements:
- pytree genotype compatibility (add the moment, CMAES and CMAME only works for jnp.ndarray)
- check random direction performance
- check CMAMEGA performance

Notes:
- About the pool of homogeneous emitters and the scheduling rule "the emitter with the less emissions emits": quite soon, only one emitter is used, because at each reinit event, it's used until it reached the others number of emission. Plus, it's not supposed to provide any difference with having just one (considering this particular emitting rule).

### CODE REVIEW

File: qdax/core/emitters/cma_emitter.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        emit_count: count the number of emission events.
+    """
+
+    random_key: RNGKey
+    cmaes_state: CMAESState
+    previous_fitnesses: Fitness
+    emit_count: int
+
+
+class CMAEmitter(Emitter):
```
Comment (Lookatator) [inline qdax/core/emitters/cma_emitter.py:38]: If I understood correctly, this corresponds to the **improvement** Emitter. 

Thus I find it weird that: (1) this class is not called CMAImprovementEmitter, and (2) the other classes inherit from it.

Wouldn't it be better to have a base (abstract) class `CMAEmitter`, with `_ranking_criteria(...)` as abstract method, and that all the other CMA[...]Emitter inherit from it?

File: qdax/core/emitters/cma_emitter.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+            This applies for the setting from the paper CMA-ME. Those facts might
+            not be true with other problems and hyperparameters.
+            """
+
+            (cmaes_state, emitter_state, repertoire, emit_count, random_key) = operand
+
+            # Update CMA Parameters
+            # mask = sorted_improvements >= 0
+            # mask = mask + 1e-6
```
Comment (Lookatator) [inline qdax/core/emitters/cma_emitter.py:257]: Why are those here? If they come from the original implementation, maybe say it in the comment on the previous line?

File: qdax/core/emitters/cma_emitter.py (hunk 1 above)
Comment (Lookatator) [reply qdax/core/emitters/cma_emitter.py:38]: Same remark for the `CMAEmitterState`

File: tests/baselines_test/cmame_test.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    def scoring_function(x: jnp.ndarray) -> Tuple[Fitness, Descriptor, Dict]:
+        scores, descriptors = fitness_scoring(x), _behavior_descriptors(x)
+        return scores, descriptors, {}
+
+    def scoring_fn(
+        x: jnp.ndarray, random_key: RNGKey
+    ) -> Tuple[Fitness, Descriptor, ExtraScores, RNGKey]:
+        fitnesses, descriptors, extra_scores = jax.vmap(scoring_function)(x)
+        return fitnesses, descriptors, extra_scores, random_key
```
Comment (Lookatator) [inline tests/baselines_test/cmame_test.py:64]: Couldn't we use directly the `rastrigin_proj_scoring_function` to make things shorter, and avoid code replication?

File: tests/core_test/cmaes_test.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    num_iterations = 10000
+    num_dimensions = 100
+    batch_size = 36
+    num_best = 36
+    sigma_g = 0.5
+    minval = -5.12
+
+    def sphere_scoring(x: jnp.ndarray) -> jnp.ndarray:
+        return -jnp.sum((x + minval * 0.4) * (x + minval * 0.4), axis=-1)
```
Comment (Lookatator) [inline tests/core_test/cmaes_test.py:20]: Same thing as above; can't we simply use the standard function defined in tasks, instead of redefining it? (if we notice a bug in the future, it might take more time to fix it in replicated codes.)

File: qdax/core/emitters/cma_emitter.py (hunk 1 above)
Comment (felixchalumeau) [reply qdax/core/emitters/cma_emitter.py:38]: I do agree for the Emitter!

For the EmitterState, the current design seems to make sense for me (as soon as the current CMAEmitter will have been renamed CMAImprovementEmitter).

File: qdax/core/emitters/cma_emitter.py (hunk 2 above)
Comment (felixchalumeau) [reply qdax/core/emitters/cma_emitter.py:257]: I moved them to the docstrings with an explanation!

File: tests/baselines_test/cmame_test.py (hunk 3 above)
Comment (felixchalumeau) [reply tests/baselines_test/cmame_test.py:64]: I perfectly agree. I thought we could keep it for a further PR that will unify all the notebooks and tests.
See issue #95 

Would you be ok to wait for another PR to do it?

Comment (Lookatator) [reply tests/baselines_test/cmame_test.py:65]: Yes I agree :)

**Review by Lookatator — APPROVED** [review]
